keps/sig-network/20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -0,0 +1,694 @@
+
+#### Kubeadm Configuration Options
+
+The kubeadm configuration options for advertiseAddress and podSubnet will need to be changed to handle a comma-separated list of CIDRs:
+```
+    api:
+      advertiseAddress: "fd00:90::2,10.90.0.2" [Multiple IP CIDRs, comma separated list of CIDRs]
+    networking:
+      podSubnet: "fd00:10:20::/72,10.20.0.0/16" [Multiple IP CIDRs, comma separated list of CIDRs]
+```
+
+#### Kubeadm-Generated Manifests
+
+Kubeadm will need to generate dual-stack CIDRs for the --service-cluster-ip-range command line argument in kube-apiserver.yaml:
+```
+    spec:
+      containers:
+      - command:
+        - kube-apiserver
+        - --service-cluster-ip-range=fd00:1234::/110,10.96.0.0/12
+```
+
+Kubeadm will also need to generate dual-stack CIDRs for the --cluster-cidr argument in kube-apiserver.yaml:
+```
+    spec:
+      containers:
+      - command:
+        - kube-controller-manager
+        - --cluster-cidr=fd00:10:20::/72,10.20.0.0/16
+```
+
+### vendor/github.com/spf13/pflag
+This dual-stack proposal will introduce a new IPNetSlice object to spf13.pflag to allow parsing of comma separated CIDRs. Refer to [https://github.com/spf13/pflag/pull/170](https://github.com/spf13/pflag/pull/170)
+
+### End-to-End Test Support
+End-to-End tests will be updated for Dual Stack. The Dual Stack E2E tests will use deployment scripts from the kubernetes-sigs/kubeadm-dind-cluster github repo to set up a containerized, multi-node Kubernetes cluster that is running in a Prow container (Docker-in-Docker-in-Docker, or DinDinD configuration), similar to the IPv6-only E2E tests (see [test-infra PR # 7529](https://github.com/kubernetes/test-infra/pull/7529)). The DinDinD cluster will be updated to support dual-stack.
+
+The E2E test suite that will be run for dual stack will be based upon the [IPv6-only test suite](https://github.com/CiscoSystems/kube-v6-test) as a baseline. New versions of the network connectivity test cases that are listed below will need to be created so that both IPv4 and IPv6 connectivity to and from a pod can be tested within the same test case. A new dual-stack test flag will be created to control when the dual stack tests are run versus single stack versions of the tests:
+```
+[It] should function for node-pod communication: udp [Conformance]
+[It] should function for node-pod communication: http [Conformance]
+[It] should function for intra-pod communication: http [Conformance]
+[It] should function for intra-pod communication: udp [Conformance]
+```
+Most service test cases do not need to be updated as the service remains single stack.
+
+For the test that checks pod internet connectivity, the IPv4 and IPv6 tests can be run individually, with the same initial configurations.
+```
+[It] should provide Internet connection for containers
+```
+
+### User Stories
+\<TBD\>
+
+### Risks and Mitigations
+\<TBD\>
+
+## Graduation Criteria
+\<TBD\>
```

Comment:
I'm going to approve this but we should start thinking about this, too.  Specifically:

* do we want to GA the multi-IP suppot (phase 1) independent of the dual-stack services work?
* what criteria will we be monitoring?

Reply:
Having dual stack on pod ip is useful for those offering ingress as ipv4 and doing downstream server calls to ipv6 (or the otherway around). I can not speak to the volume of users. I think as we release phase 1 to alpha we will hear from users and we can adjust accordingly.

@lachie83